refactor(connected-overlay-directive): avoid issues with property initialization

crisbeto opened the pull request:
* Refactors the `ConnectedOverlayDirective` not to depend on the order in which the `open` and `origin` properties are set.
* Moves the `open`, `offsetX` and `offsetY` properties from using setters to `ngOnChanges` since the latter tends to output less JS.

Fixes #3225.

kara commented:
@crisbeto This appears to break md-select positioning. Can you fix?

crisbeto commented:
Fixed @kara. I don't know if calling `detectChanges` is considered a code smell, but it should be shorter than getters/setters. If you think it may be problematic, I'll switch it back.

jelbourn commented on src/lib/core/overlay/overlay-directives.ts:
```diff
@@ -172,6 +144,20 @@ export class ConnectedOverlayDirective implements OnDestroy {
 
+  ngOnChanges(changes: SimpleChanges) {
+    if (changes['open']) {
+      changes['open'].currentValue ? this._attachOverlay() : this._detachOverlay();
```
After you check `changes['open']`, shouldn't you be able to just use `this.open`?
(same for others)

jelbourn commented on src/lib/select/select.ts:
```diff
@@ -577,6 +577,7 @@ export class MdSelect implements AfterContentInit, ControlValueAccessor, OnDestr
     }
 
     this._checkOverlayWithinViewport(maxScroll);
+    this._changeDetectorRef.detectChanges();
```
Add comment explaining why this `detectChanges` call is necessary

jelbourn commented on src/lib/core/overlay/overlay-directives.ts:
```diff
@@ -76,30 +75,10 @@ export class ConnectedOverlayDirective implements OnDestroy {
-      this._position.withOffsetX(offsetX);
-    }
-  }
+  @Input() offsetX: number = 0;
```
How significant is the code size difference? We use quite a lot of getter/setters.

crisbeto replied:
A getter/setter combination transpiles to this in ES5:

```
Object.defineProperty(A.prototype, "thing", {
    get: function () {
        return this._thing;
    },
    set: function (value) {
        this._thing = value;
    },
    enumerable: true,
    configurable: true
});
```

I think that it can definitely start adding up over an entire project.

crisbeto commented:
Addressed the feedback @jelbourn.

crisbeto commented:
@jelbourn I've rebased this one and removed the changes to the `offsetX` and `offsetY` since they also turned out to be a bit dangerous. Can you take another look?

jelbourn commented on src/lib/core/overlay/overlay-directives.ts:
```diff
@@ -183,6 +177,12 @@ export class ConnectedOverlayDirective implements OnDestroy {
+      this.open ? this._attachOverlay() : this._detachOverlay();
+    }
+  }
+
```
Can you add a unit test for the case this fixes?

crisbeto commented:
Added a test case for the changes @jelbourn.
